src/librustc_mir/borrow_check/nll/region_infer/error_reporting/mod.rs

```diff
@@ -465,9 +468,93 @@ impl<'tcx> RegionInferenceContext<'tcx> {
     }
 
+    fn add_static_impl_trait_suggestion(
+        &self,
+        infcx: &InferCtxt<'_, '_, 'tcx>,
+        diag: &mut DiagnosticBuilder<'_>,
+        fr: RegionVid,
+        // We need to pass `fr_name` - computing it again will label it twice.
+        fr_name: RegionName,
+        outlived_fr: RegionVid,
+    ) {
+        if let (
+            Some(f),
+            Some(ty::RegionKind::ReStatic)
+        ) = (self.to_error_region(fr), self.to_error_region(outlived_fr)) {
+            if let Some(ty::TyS {
+                sty: ty::TyKind::Anon(did, substs),
+                ..
+            }) = infcx.tcx.is_suitable_region(f)
+                    .map(|r| r.def_id)
+                    .map(|id| infcx.tcx.return_type_impl_trait(id))
+                    .unwrap_or(None)
+            {
+                // Check whether or not the impl trait return type is intended to capture
+                // data with the static lifetime.
+                //
+                // eg. check for `impl Trait + 'static` instead of `impl Trait`.
+                let has_static_predicate = {
+                    let predicates_of = infcx.tcx.predicates_of(*did);
+                    let bounds = predicates_of.instantiate(infcx.tcx, substs);
+
+                    let mut found = false;
+                    for predicate in bounds.predicates {
+                        if let ty::Predicate::TypeOutlives(binder) = predicate {
+                            if let ty::OutlivesPredicate(
+                                _,
+                                ty::RegionKind::ReStatic
+                            ) = binder.skip_binder() {
+                                found = true;
+                                break;
+                            }
+                        }
+                    }
+
+                    found
+                };
+
+                debug!("add_static_impl_trait_suggestion: has_static_predicate={:?}",
+                       has_static_predicate);
+                let static_str = keywords::StaticLifetime.name();
+                // If there is a static predicate, then the only sensible suggestion is to replace
+                // fr with `'static`.
+                if has_static_predicate {
+                    diag.help(
+                        &format!(
+                            "consider replacing `{}` with `{}`",
+                            fr_name, static_str,
+                        ),
+                    );
+                } else {
+                    // Otherwise, we should suggest adding a constraint on the return type.
+                    let span = infcx.tcx.def_span(*did);
+                    if let Ok(snippet) = infcx.tcx.sess.source_map().span_to_snippet(span) {
+                        let suggestable_fr_name = match fr_name {
+                            RegionName::Named(name) => format!("{}", name),
+                            RegionName::Synthesized(_) => "'_".to_string(),
```

Comment:
Hmm, I just noticed this; it is not always going to be correct, right? 

e.g., in this example

```rust
fn foo(x: &[i32], y: &[i32]) -> impl Iterator<Item = i32> {
    y.iter().cloned()
}
```

but we should probably just file a follow-up issue.

Reply:
We currently get the following for that example:

```
error[E0621]: explicit lifetime required in the type of `y`
 --> test.rs:2:5
  |
1 | fn foo(x: &[i32], y: &[i32]) -> impl Iterator<Item = i32> {
  |                      ------ help: add explicit lifetime `'static` to the type of `y`: `&'static [i32]`
2 |     y.iter().cloned()
  |     ^^^^^^^^^^^^^^^^^ lifetime `'static` required

error: unsatisfied lifetime constraints
 --> test.rs:2:5
  |
1 | fn foo(x: &[i32], y: &[i32]) -> impl Iterator<Item = i32> {
  |           -          - let's call the lifetime of this reference `'1`
  |           |
  |           let's call the lifetime of this reference `'2`
2 |     y.iter().cloned()
| ^^^^^^^^ requires that `'1` must outlive `'2`
```